Add Proposal for dependency resolution in kc

praveenrewar opened the pull request.

praveenrewar committed:
Add Proposal for dependency resolution in kc

Signed-off-by: Praveen Rewar <8457124+praveenrewar@users.noreply.github.com>

everettraven commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,621 @@
+## Proposal
+In this proposal we will describe the changes that are necessary to enable the Packaging API to support dependency management.
+
+With these changes we
```
Unfinished sentence or lingering artifact from a rewording?

varshaprasad96 commented on proposals/kapp-controller/004-dependency-resolution/README.md:
````diff
@@ -0,0 +1,621 @@
+  -> pkg3
+```
+
+For future versions we can contemplate the possibility of each PackageInstall checking if the dependencies it has can be fulfilled with the Packages present in the Cluster.
````
This means the proposal does not cover runtime dependencies. If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)?

praveenrewar replied:
This sentence should have been removed (my bad). The following condition in the status already covers this scenario where if an already installed package satisfies the dependency then we don't install the package again
```yaml   
- name: microservice-v1
      wantedPackageRef:
        refName: pkg5.test.carvel.dev
        versionSelection:
          constraints: 0.2.0
      packageInstallRef:
        name: other-pkg.pkg5.test.carvel.dev
        namespace: ns1
      conditions:
      - type: FoundPackageInstall
 ```

praveenrewar replied:
> If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)

We want to add a `installationType` field on the package for this (check package crd). This will determine if the package can be installed again or not. If it cannot be installed again and the version installed on the cluster doesn't satisfy the version constraints provided on the package for the dependency, then we will fail with 
```yaml
    - name: dep-with-wrong-version
      wantedPackageRef:
        refName: pkg1.test.carvel.dev
        versionSelection:
          constraints: "1.0.1" # one on the cluster is 0.1.1
      packageInstallRef:
        name: pkg-demo.pkg1.test.carvel.dev
      conditions:
      - type: FoundPackageInstall
      - type: FoundPackageInstallRequiresDifferentVersion
  ```

prembhaskal replied:
> If pkg 1 --> pkg 2 (v1.0.0), and the cluster already contains pkg2 (v1.1.0) installed - do we just fail (even though it is just a minor bump and is expected to be backwards compatible)
@praveenrewar  since we support range in the versionSelection.constraints, then I think users can get this behaviour if they choose to.

varshaprasad96 commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,619 @@
+##### Package depends on another Package
+- When the Package that match the dependency is already installed, no new installation is needed
+- When the dependent Pacakage is not installed, it will try to install it
+- When a dependent Package is already installed but the version does not fullfil the requirements of the current PackageInstall, kapp-controller will try to install the version that matches
```
What happens to the older version, its upgraded or both versions are installed? It would be helpful to add some more thoughts around that (Given I assume we don't different versions of the same GK owned by multiple packages, or is this not the case?)

praveenrewar replied:
kapp-controller will try to install the version that is required by this package, the existing installation will remain as is. I will add more details around this, thanks!

varshaprasad96 commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,619 @@
+- When there is no Package that fullfills the dependency of a PackageInstall the installation will fail
+
+##### Two Packages depend on the same Package
+In this scenario the first PackageInstall that creates the dependent PackageInstall will own it. This means that the second PackageInstall will not need to create the dependent PackageInstall
```
How does this work with transitive dependencies? Pkg A -> pkg B - > pkg C (with `->` signifying one depends on another)? I assume pkg C would be owned by pkg B (but it would be nice to add some info on the same).

praveenrewar replied:
Sure, I will add more info around it. The current plan is to only deal with first level of dependency in the first iteration and then think about transitive dependencies in further iterations. 
However, even with one level of dependency, since by default kapp-controller will try to install the dependencies, even pkg C would be installed, and yes, it would be owned by pkg B (will add details around it).

varshaprasad96 commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,619 @@
+In this scenario the first PackageInstall that creates the dependent PackageInstall will own it. This means that the second PackageInstall will not need to create the dependent PackageInstall
+
+#### Deletion of PackageInstalls
+When a PackageInstall which was responsible for installing one or more dependencies is removed, we will just orphan the dependent package installs (i.e remove the "kapp-controller.carvel.dev/owner" annotation) and not delete them.
```
How would garbbage collection of packageinstalls work? Is there going to be a separate controller to handle the same?

praveenrewar replied:
> How would garbage collection of packageinstalls work?

Do you mean garbage collection of the dependencies?
For the first iteration we are not going to do garbage collection. In further iterations, we will allow users to decide if they want the dependencies to be gone as well (we can add a check to make sure that no other package is dependent on these dependencies).

> the scenario may occur where the orphaned package existing on cluster is requested for installation.

Do you mean an orphaned dependency is now required by another package install? (The above scenario I mentioned should cover this I guess)

varshaprasad96 reviewed (commented):
@praveenrewar Thanks for the proposal. This is a very useful feature to have in kapp-controller. Just a few more questions:
1. Is the proposal intended to cover the design of the resolver itself? (In OLM we do use a [SAT solver](https://github.com/operator-framework/deppy), so just curious if there is some investigation in progress to figure out the options).
2. Do users creating package install have the option of doing a force install?
3. Given the dependent package installs are exposed to the user, what happens when it is deleted directly? In the sense, do we have watches set up on dependent packages and how frequently does the resolution occur?

rashmigottipati commented on proposals/kapp-controller/004-dependency-resolution/README.md:
````diff
@@ -0,0 +1,619 @@
+
+ex:
+```
+When pkg depends on pkg1 and pkg3, while pkg1 depends on pkg2
````
Would pkg2 still be captured in pkg even though it is not directly responsible for installing it? Because if pkg1 is deleted, then it will cause a chain reaction of failures for the packages that rely on pkg2.

praveenrewar replied:
`pkg2` won't be captured in `pkg`, it will be captured in `pkg1` only. 
If the dependencies of `pkg1` are deleted, then in the further reconciliations of `pkg1` kapp-controller will try to install these dependencies again, but, yes, during that period of time, there might be failures in both `pkg1` and `pkg`. In future we might think about adding a _watch_  on dependencies but for now we are going to rely on kapp-controller's reconciliations.

rashmigottipati commented on proposals/kapp-controller/004-dependency-resolution/README.md:
````diff
@@ -0,0 +1,619 @@
+```
+
+#### Dependency resolution
+The dependency resolution will be done one PackageInstall at a time, basically each PackageInstall will only create the PackageInstalls that it directly depends on.
````
When there is a new dependency that gets identified for a specific `PackageInstall`, can the `PackageInstall` be updated with that dep after it's been installed in the cluster?

praveenrewar replied:
Yes! kapp-controller will try to reconcile PackageInstalls after every `syncPeriod` duration, so in the meanwhile if there have been any new dependencies added to the package, then those would be installed in the next reconciliation.

praveenrewar commented:
@varshaprasad96 
> Is the proposal intended to cover the design of the resolver itself? 

We currently use a resolver based on the semver constraints in the package install api to resolve the versions, we are planning to re-use it for dependency resolution. (Does this answer your question?)

> Do users creating package install have the option of doing a force install?

Do you mean if users can install packages without installing the dependencies? Then, yes, users can set the `spec.dependencies.managedBy` field to `User` (default is `Auto`). However, kc would try to validate if the dependent packages (not package installs) are present on the cluster or not.

> Given the dependent package installs are exposed to the user, what happens when it is deleted directly? In the sense, do we have watches set up on dependent packages and how frequently does the resolution occur?

In the first few iterations at least, we won't have an watches set for the dependencies, but we could think about it in future enhancements. This is a good point, I will add it to the list of future enhancements. Thanks!

vrabbi commented:
comments/questions:

 The doc contradicts itself. In the crd spec you have User as the default mode but in the example the comment says the default will be Auto in terms of dependency installation. Which is the approach you are planning on implementing?

 What if packages are not in the same namespace? Does this require them to all be in the same namespace and or be in the global namespace?

 Do all dependencies use the same service account as the main package install or is a service account able to be specified for different dependencies?

 How can overlays be applied to dependant packages? Would this require manual dependency installation?

 Does this work for targetting a different cluster as well or only for in cluster package installations?

 Would canceling/pausing the main packagr install have an impact and cascade to dependencies or not?

 Are changes made manually to auto installed dependent package installs reverted at next reconcile? Can one manually pause/cancel reconcilliation of a dependency?

 Is the sync period of the dependencies copied from the main app or always set to the default value?

 The doc says that when a package already is installed but not of the version which is needed for the new package install, kapp controller will try to install the new version, does this mean it will try to update the version of the existing package install or create another one? If its upgrading the existing one, how does kapp controller know the newer version is safe for the other root packages depending on the original dependency version? If it tries to install another pkgi side by side, what if the install type is once per cluster or once per namespace?

 The spec of the package CRD is not inline with the example. Thr example has a packageRef section with version constraints which is great whereas the CRD spec has a version field only under a package key which seems to not support version ranges and constraints. Which approach is planned to be implemented?

praveenrewar commented:
Thanks a lot @vrabbi!! These are some really great questions! I will answer them inline here and then try to move them to the proposal (either update it or add the question in the "Answered questions sections).

> In the crd spec you have User as the default mode but in the example the comment says the default will be Auto in terms of dependency installation. 

Good catch! `Auto` will be the default value. I have updated it in the proposal.

> What if packages are not in the same namespace? Does this require them to all be in the same namespace and or be in the global namespace?

For now the packages should be in the same namespace, in future we can think about expanding the search to the whole cluster, this will be important for packages which have installationType as `OncePerCluster`.

> Do all dependencies use the same service account as the main package install or is a service account able to be specified for different dependencies?

Yes, for now, the dependencies will be using the same service account. In future, we can think of alternatives where a different SA is used for each dependency.

> How can overlays be applied to dependant packages? Would this require manual dependency installation?

I feel that in this case we reach a stage where we are trying to control a lot of stuff on the dependency so should be manually installing the dependency, but if you think that we should allow adding overlays for dependencies directly, then we can think about having something similar to the data values for the overlays as well.

> Does this work for targetting a different cluster as well or only for in cluster package installations?

In theory, it should work when you are targeting another cluster. You should have the dependency packages available on the cluster you are creating the current package install, and then if kapp-controller tries to use the same kubeconfig secret for installing the dependency packages on the target cluster, it should work.

> Would canceling/pausing the main packagr install have an impact and cascade to dependencies or not?

There won't be any cascading of the `paused` field. Do you think we should pause the dependencies as well when the original pkgi is paused?

> Are changes made manually to auto installed dependent package installs reverted at next reconcile? Can one manually pause/cancel reconcilliation of a dependency?

Yes, the changes would be reverted on the next reconciliation. We can pause both the original pkgi and the dependency pkgi if we want. 

> If it tries to install another pkgi side by side, what if the install type is once per cluster or once per namespace?

For this initial iterations, kapp-controller will fail when a certain version of a dependency is installed which doesn't satisfy the constraints provided by our package (one of the conditions in the example covers this). In future versions, we will try to install another version of the package if the installationType is `NoRestrictions`, if it is `OncePerCluster` or `OncePerNamespace` then we will fail with the same condition.

> The spec of the package CRD is not inline with the example

Thanks. I have updated it. We will use packageRef.

vrabbi commented:
Thanks for the detailed response. 

I think overlays are important and in many cases are needed so having that option makes sense to me and i have a few use cases where this would make a lot of sense. 

In terms of pausing dependencies i think its fine that it doesnt pause them it just needs to be clear in the docs/proposal. In an ideal world i could see value in having a paused field under each dependency in the main package that would be how to pause a dependency. This is better than pausing the dependency directly as all management of the dependency is through a single resource vs it being from multiple sources. Also it would work better for gitops solutions where only the main package is controlled by git and as such would allow for more advanced use cases. If this were done, i would expect changes done directly on the dependencies including setting the paused field to be overriden as the source of truth is the main packageinstall

Another new question is how could one "orphan" a dependency that was auto installed so it could be managed seperately. I can think of many cases where initially the user wants to manage a dependency simply via thr high level package but times may come when they want to manage it seperately as they have more specific use cases they want to have configurable. In this case i think that breaking the ownership would be a needed ability for users.

prembhaskal commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,628 @@
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
```
can we have multiple owners for a package install for cases like
pkg1 brought pkg3
pkg2 needs pkg3, but it is already satisfied.
in the 2nd case, do we add the pkg2 also as owner.

prembhaskal replied:
I was thinking about how to handle the delete case when one package with multiple parent pkgi gets orphaned. but it is ok, we can manage it during implementation.

prembhaskal commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,628 @@
+In this scenario the first PackageInstall that creates the dependent PackageInstall will own it. This means that the second PackageInstall will not need to create the dependent PackageInstall
+
+#### Deletion of PackageInstalls
+When a PackageInstall which was responsible for installing one or more dependencies is removed, we will just orphan the dependent package installs (i.e remove the "kapp-controller.carvel.dev/owner" annotation) and not delete them.
```
Similar question as above. 
Assuming we support only single owner on a packageinstall
t0 ==> pkg1 -> pkg3 so pkg3 has owner=pkg1
t1  ==> pkg2 -> pkg3 AlreadySatisfied dependency, so we don't update owner
...
t2 ==> pkg1 is deleted, for pkg3 do we now remove owner=pkg1 or change it to owner=pkg3

prembhaskal replied:
yes, i meant pkg2 (edited now).

praveenrewar commented:
Thanks @vrabbi for the thoughts.

We can include overlays for dependencies as well, maybe in future iterations as we are evaluating if we should have a overlays as a first class api in the package install instead of annotations.

Makes sense regarding not pausing the dependencies, I will update the proposal to make it more clear.
I like the gitops idea, I will try to capture it in the future explorations section. It sounds good to control the dependencies from one source, we will have to think more about the scenarios where there the dependencies are also shared by other packages.

 > Another new question is how could one "orphan" a dependency that was auto installed so it could be managed seperately.
 
 Hmm, definitely an interesting use case. I am thinking that from the current implementation perspective, removing the owner annotation manually from the dependency would be one way, although I think it would be useful to have a way to de-link the dependency from the owner pkgi. I will add this in the future explorations section as well.

joaopapereira commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,639 @@
+
+## Open Questions
+
+- What happens if the Resource did not change version but new fields were added and the package requires them but that version of the Resource is not present in the cluster?
```
```suggestion
```

joaopapereira reviewed (approved):
I think we are in a good place with this proposal. We can remove the Open Question since it is already answered. 
Let us also try to close all the open conversations, and if no outstanding comments are added I think we can mark it as approved somewhere middle next week.

praveenrewar commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,637 @@
+metadata:
+  name: pkg1.test.carvel.dev.1.0.1
+  annotations:
+    kapp-controller.carvel.dev/owner: "PackageInstall/pkg-demo"
```
Switched to `kapp-controller.carvel.dev/package-install-owner: "pkg-demo"`.

praveenrewar commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,637 @@
+  version: 1.0.0
+  dependencies:
+    - gvk:
+        group: api.carvel.dev
```
Intention was to use `apiVersion` and `kind`, this was probably an old example which I missed.

praveenrewar commented on proposals/kapp-controller/004-dependency-resolution/README.md:
```diff
@@ -0,0 +1,637 @@
+app_port: 8080
+hello_msg: stranger
+
+---
+# dep=dep-1               # key = dep, value = name of dependency
+svc_port: 80
```
Agree with the rationale. I have removed this section for now, we can revisit this in the future when we have more demand for this.

praveenrewar commented:
@cppforlife I have addressed your comments. Please give it another look when you get a chance.

joaopapereira commented:
@praveenrewar, please create a new commit with the change of status to approved since we have already passed the consensus period.

praveenrewar committed:
Mark proposal as approved

Signed-off-by: Praveen Rewar <8457124+praveenrewar@users.noreply.github.com>

praveenrewar commented:
@joaopapereira Done!
